dali/pipeline/data/tensor_vector.cc

```diff
@@ -766,22 +755,81 @@ TensorVector<Backend> &TensorVector<Backend>::operator=(TensorVector<Backend> &&
     tl_ = std::move(other.tl_);
     type_ = other.type_;
     sample_dim_ = other.sample_dim_;
-    views_count_ = other.views_count_.load();
     tensors_ = std::move(other.tensors_);
-    for (auto &t : tensors_) {
-      if (t) {
-        if (auto *del = std::get_deleter<ViewRefDeleter>(t->data_)) del->ref = &views_count_;
-      }
-    }
 
-    other.views_count_ = 0;
     other.curr_num_tensors_ = 0;
     other.tensors_.clear();
   }
   return *this;
 }
 
 
+// This is to check if we are actually laid down in contiguous memory
+template <typename Backend>
+bool TensorVector<Backend>::IsContiguousInMemory() const {
+  if (num_samples() == 0 || shape().num_elements() == 0) {
+    return true;
+  }
+  // If we are using contiguous representation in this case we can safely return
+  if (IsContiguous()) {
+    return true;
+  }
+  const uint8_t *base_ptr = static_cast<const uint8_t *>(tensors_[0]->raw_data());
+  size_t size = type_info().size();
+
+  for (int i = 0; i < num_samples(); ++i) {
+    if (base_ptr != tensors_[i]->raw_data()) {
+      return false;
+    }
+    base_ptr += tensors_[i]->shape().num_elements() * size;
+  }
+  return true;
+}
+
+
+template <typename Backend>
+bool TensorVector<Backend>::IsDenseTensor() const {
+  return IsContiguousInMemory() && is_uniform(shape());
+}
+
+
+template <typename Backend>
+Tensor<Backend> TensorVector<Backend>::AsReshapedTensor(const TensorShape<> &new_shape) {
+  DALI_ENFORCE(new_shape.num_elements() == 0 || num_samples() > 0,
+               "To create a non-empty view Tensor, the batch must not be empty.");
+  DALI_ENFORCE(IsValidType(type()),
+               "To create a view Tensor, the batch must have a valid data type.");
+  DALI_ENFORCE(
+      shape().num_elements() == new_shape.num_elements(),
+      make_string("To create a view Tensor, requested shape need to have the same volume as the "
+                  "batch, requested: ",
+                  new_shape.num_elements(), " expected: ", shape().num_elements()));
+  Tensor<Backend> result;
+  result.ShareData(unsafe_owner(*tl_), tl_->capacity(), tl_->is_pinned(), new_shape, type(),
+                   device_id(), order());
```

Comment:
Now it requires contiguous allocation, not just incidental contiguity :(
```suggestion
  shared_ptr<void> ptr = num_samples() > 0 ? unsafes_sample_owner(*tl_, 0) : IsContiguous() ? unsafe_owner(*tl_) : nullptr;
  result.ShareData(ptr, tl_->capacity(), tl_->is_pinned(), new_shape, type(),
                   device_id(), order());
```

Reply:
Now it is even worse, as we are taking some parameters like capacity from the contiguous allocation and some from the accidentally contiguous case.

Reply:
And I have conscientious objection to nested ternary.

Reply:
done